st2common/st2common/services/trace.py

```diff
@@ -0,0 +1,258 @@
+        parent_execution_id = liveaction.context['parent'].get('execution_id', None)
+        if parent_execution_id:
+            # go straight to a trace_db. If there is a parent execution then that must
+            # be associated with a Trace.
+            trace_db = get_trace_db_by_action_execution(action_execution_id=parent_execution_id)
+            if not trace_db:
+                raise StackStormDBObjectNotFoundError('No trace found for execution %s' %
+                                                      parent_execution_id)
+            return trace_db
+    # 3. Check if the action_execution associated with liveaction leads to a trace_db
+    execution = ActionExecution.get(liveaction__id=str(liveaction.id))
+    if execution:
+        trace_db = get_trace_db_by_action_execution(action_execution=execution)
+    # 4. No trace_db found, therefore create one. This typically happens
+    #    when execution is run by hand.
+    if not trace_db:
+        trace_db = TraceDB(trace_id='execution-%s' % str(liveaction.id))
+    return trace_db
+
+
+def add_or_update_given_trace_context(trace_context, action_executions=None, rules=None,
+                                      trigger_instances=None):
+    """
+    Will update an existing Trace or add a new Trace. This method will only look for exact
+    Trace as identified by the trace_context. Even if the trace_context contain a trace_id
+    it shall not be used to lookup a Trace.
+
+    * If an exact matching Trace is not found a new Trace is created
+    * Whenever only a trace_id is supplied a new Trace is created.
+
+    :param trace_context: context object using which a trace can be found. If not found
+                          trace_context.trace_id is used to start new trace.
+    :type trace_context: ``dict`` or ``TraceContext``
+
+    :param action_executions: The action_execution to be added to the Trace. Should a list
+                              of object_ids.
+    :type action_executions: ``list``
+
+    :param rules: The rules to be added to the Trace. Should a list of object_ids.
+    :type rules: ``list``
+
+    :param trigger_instances: The trigger_instances to be added to the Trace. Should a list
+                              of object_ids.
+    :type trigger_instances: ``list``
+
+    :rtype: ``TraceDB``
+    """
+    trace_db = get_trace(trace_context=trace_context, ignore_trace_id=True)
+    if not trace_db:
+        # since trace_db is None need to end up with a valid trace_context
+        trace_context = _get_valid_trace_context(trace_context)
+        trace_db = TraceDB(trace_id=trace_context.trace_id)
+    return add_or_update_given_trace_db(trace_db=trace_db,
+                                        action_executions=action_executions,
+                                        rules=rules,
+                                        trigger_instances=trigger_instances)
+
+
+def add_or_update_given_trace_db(trace_db, action_executions=None, rules=None,
```

Comment:
Yeah I have definitely left that part a little open right now. I was thinking we sort client side since it will be a merge sort anyway.